### PR TITLE
Replace refer all

### PR DESCRIPTION
**Pull request by benedekfazekas**
Replace :refer :all style require with alias style require or
optionally (with prefix) referred names style require.

- with prefix cljr--replace-refer-all-with-referred-list still available
- add keybinding for cljr-replace-refer-all
- add cljr-replace-refer-all to the ns hydras
- add warm-macro-occurrences-cache
- defcustom to warm-macro-occurrences-cache at startup

clojure-emacs/refactor-nrepl#135 provides the middleware backend function

Intended to be merged for version 2.1.0

### CODE REVIEW

**Comment by expez** [comment]
We already have `cljr-stop-referring`, but it only works on aliased stuff and doesn't handle `:refer :all`.  Perhaps this new code should just extend `cljr-stop-referring` instead of being its own command?

File: clj-refactor.el [hunk 1]
```diff
@@ -158,6 +158,12 @@ as can be."
   :group 'cljr
   :type 'boolean)
 
+(defcustom cljr-eagerly-cache-macro-occurrences-on-startup t
+  "If t, the middleware will eagerly populate the macro occurrences cache.
+This makes `cljr-replace-refer-all' as it can be."
```
Comment (expez) [inline clj-refactor.el]: Missing word here

File: clj-refactor.el [hunk 2]
```diff
@@ -1372,6 +1381,38 @@ returns (used.ns.lib1 used.ns.lib2)"
...
+  (when (search-forward ":require" nil t)
+    (let (nses
+          (require-end (save-excursion
+                         (progn
+                           (paredit-backward-up)
+                           (paredit-forward)
+                           (point)))))
+      (if (not (save-excursion (re-search-forward ":refer\s:all" require-end t)))
+          (message "There is no :refer :all in the ns declaration.")
+        (while (save-excursion (re-search-forward ":refer\s:all" require-end t))
```
Comment (expez) [inline clj-refactor.el]: re-search-forward both moves point and returns a truthy value when it succeeds, so this while loop can be simplified slightly.

File: clj-refactor.el [hunk 3]
```diff
@@ -1372,6 +1381,38 @@ returns (used.ns.lib1 used.ns.lib2)"
...
+          (re-search-forward ":refer\s:all" require-end t)
+          (paredit-backward-up)
+          (let ((prefix (save-excursion
+                          (when (progn
+                                  (paredit-backward-up)
+                                  (not (looking-at "(:require")))
```
Comment (expez) [inline clj-refactor.el]: prefer `looking-at-p` here

File: clj-refactor.el [hunk 4]
```diff
@@ -2981,8 +3028,8 @@ See: https://github.com/clojure-emacs/clj-refactor.el/wiki/cljr-find-usages"
         (when (looking-at-p "(\\s-*def")
           (re-search-forward name)
           (paredit-backward))
-        (when (looking-at (format "\\(.+/\\)?\\(%s\\)" name))
-          (replace-match (format "\\1%s" new-name))))
+        (when (re-search-forward (format "\\(.+/\\)?\\(%s\\)" name) search-limit t)
```
Comment (expez) [inline clj-refactor.el]: This is what we had back in the day, but we had to change it to the current version because it caused problems.  I think this might introduce a regression.  You should walk the git history to find out exactly why we changed it the last time.

File: clj-refactor.el [hunk 5]
```diff
@@ -3017,6 +3069,51 @@ See: https://github.com/clojure-emacs/clj-refactor.el/wiki/cljr-rename-symbol"
...
+                         (--map (gethash :name it) publics-occurrences)))))))
+
+(defun cljr--replace-refer-all-with-alias (ns publics-occurrences alias)
+  (let ((ns-last-token (-last-item (s-split "\\." ns))))
+    (when (re-search-forward (format "\\(%s\s\\):refer\s:all" ns-last-token) nil t)
+      (replace-match (format "\\1:as %s" alias)))
+    (cljr--rename-occurrences ns publics-occurrences (lambda (old-name) (concat alias "/" old-name)))))
+
+;;;###autoload
+(defun cljr-replace-refer-all (with-referred-list)
```
Comment (expez) [inline clj-refactor.el]: Perhaps a better name here would `replace-referred-symbols?`

File: clj-refactor.el [hunk 6]
```diff
@@ -3017,6 +3069,51 @@ See: https://github.com/clojure-emacs/clj-refactor.el/wiki/cljr-rename-symbol"
...
+  "Replaces :refer :all style require with alias :as style require.
+    Also adds the alias prefix to all occurrences of public symbols
```
Comment (expez) [inline clj-refactor.el]: Indentation here looks a bit weird

File: features/replace-refer-all.feature [hunk 7]
```diff
@@ -0,0 +1,52 @@
+Feature: replace refer all with referred list or aliases
+
+  Background:
+    Given I switch project-clean-prompt off
```
Comment (expez) [inline features/replace-refer-all.feature]: This seems irrelevant

File: clj-refactor.el (hunk 4 above)
Comment (benedekfazekas) [inline clj-refactor.el]: hm... I think what you mean is #157 as renaming `arg` in this works fine:

```clj
(defn test [arg]
  (println arg arg arg))
 ```

I guess there is no regression here.

Comment (expez) [inline clj-refactor.el]: I think this might manifest when you're trying to rename `foo` when `foobar` also exists earlier on the same line.

Comment (benedekfazekas) [inline clj-refactor.el]: tested that as well, renaming foo works fine as well:

```clj
(defn testfoo [foobar foo]
  (println foobar foo))
```

hit me with some failing testcase if you still think there is a regression here pls.

Comment (expez) [inline clj-refactor.el]: > tested that as well, renaming foo works fine as well

:+1:

**Comment by expez** [comment]
I still think we should've extended `cljr-stop-referring` like I suggested above.  Why make it more difficult on ourselves and our users?  I'm already having a hard time remembering all the keybindings and I use this package every day.

File: clj-refactor.el [hunk 8]
```diff
@@ -149,6 +149,12 @@ as can be."
   :group 'cljr
   :type 'boolean)
 
+(defcustom cljr-eagerly-cache-macro-occurrences-on-startup t
```
Comment (expez) [inline clj-refactor.el]: Why make this a defcustom?  I don't see any reason to ever turn this off.  Do you?

**Comment by benedekfazekas** [comment]
> I still think we should've extended `cljr-stop-referring`

I am not against it. will deffo (re)consider

File: clj-refactor.el (hunk 8 above)
Comment (benedekfazekas) [inline clj-refactor.el]: can take a long time and eats quite a bit of CPU at REPL startup. some users might not want this for some reason...
